### DISCUSSION TITLE
Three.js rendering

### DISCUSSION THREAD

**Discussion by hiiambob89** [phetsims/community#25] [Q&A] [answered]
So I've been trying to get threejs working for rendering in my project but it isnt rendering on the screen.

This is a snippet that is located in my sim ScreenView constructor:
`
    const testMesh = new THREE.Mesh( new THREE.BoxGeometry(), new THREE.MeshBasicMaterial());
    const testGroup = new THREE.Group();
    testGroup.add(testMesh)
    const scene = new THREE.Scene()
    scene.add(testGroup)
    const camera = new THREE.PerspectiveCamera(75, this.layoutBounds.maxX / this.layoutBounds.maxY, 0.1, 1000);
    scene.add(camera)
    const geometry = new THREE.BoxGeometry( 1, 1, 1 );
    const material = new THREE.MeshBasicMaterial( { color: 0x00ff00 } );
    const cube = new THREE.Mesh( geometry, material );
    scene.add( cube );
    camera.position.z = 5;
    this.sceneNode = new ThreeIsometricNode(this.layoutBounds)
    this.sceneNode.stage.threeScene.add(testGroup)
    this.addChild(this.sceneNode)
`
I saw a render function called in another simulation called like this: `this.sceneNode.render(undefined);`

I must be missing something because I cant seem to get anything to render on the screen.

What should I do to get something to render?

**Comment by jonathanolson** [comment]
Does your code exist somewhere where I could try it out and reproduce?

`sceneNode.render( undefined )` should be called to update the view (within the Display). The ThreeIsometricNode's stage actually already has a camera, so it would use that one (and ignore the camera that was created).

The most minimal example I can provide right now is https://github.com/phetsims/mobius/blob/e1765f1f10b26c84cc2cc4ab715ddf4c9cbd6170/js/MobiusDemoScreenView.ts, which should look something like:

<img width="522" alt="image" src="https://github.com/phetsims/community/assets/480878/27c7ab0e-d77b-4f74-841a-ded141f15c4c">

Note that ScreenView's `step()` is called every frame, and renders an updated view (https://github.com/phetsims/mobius/blob/e1765f1f10b26c84cc2cc4ab715ddf4c9cbd6170/js/MobiusDemoScreenView.ts#L110).

**Reply by hiiambob89** [reply]
https://github.com/hiiambob89/Phet-Dynamics-Lab-Bead-On-Hoop
Here is a link, I copied the demo code and imports into my project for a moment just to see if I could reproduce it. I get a WebGL is not enabled or available error. When running the actual mobius demo, however, it works just fine.

**Reply by samreid** [reply, marked as answer]
Thanks, I cloned your repo into a directory "bead-on-hoop/" then commented out simData.js and saw this error:

![image](https://github.com/phetsims/community/assets/679486/8c1fb023-3df9-4ad0-9792-ac28fc01eb71)

Then I added a query parameter ?webgl=true and it rendered:

![image](https://github.com/phetsims/community/assets/679486/858bb3d7-1e13-447b-a0c4-60605811863c)

It looks like the same thing can be accomplished by putting `webgl: true` in the SimOptions, as was done in https://github.com/phetsims/mobius/blob/e1dcfc5d9b14c6d950cd8caf05d631e8547a0fbe/js/mobius-main.ts#L28.  I tested that in bead-on-hoop/ and it worked well (so it will run without the query parameter.)

Our apologies for the lack of documentation and clarity around this part!

**Reply by hiiambob89** [reply]
Thank You!

**Reply by hiiambob89** [reply]
So doing it as a query parameter does work, but adding that line into SimOptions didn't work on Chrome for me. It worked on Firefox, but not on Chrome.

**Reply by jessegreenberg** [reply]
Hi @hiiambob89 - I just checked out your repository at https://github.com/hiiambob89/Phet-Dynamics-Lab-Bead-On-Hoop and applied this and I am able to see the rotating cube in Chrome.

```patch
Subject: [PATCH] Initial commit for code creation
---
Index: js/bead-on-hoop-main.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/bead-on-hoop-main.ts b/js/bead-on-hoop-main.ts
--- a/js/bead-on-hoop-main.ts	(revision a1571292c63045fbe4957d448270b227a55ff14e)
+++ b/js/bead-on-hoop-main.ts	(date 1690839893500)
@@ -25,6 +25,8 @@
 
   const options: SimOptions = {
 
+    webgl: true,
+
     //TODO fill in credits, all of these fields are optional, see joist.CreditsNode
     credits: {
       leadDesign: '',
```

Chrome will often cache changes unless you disable it, can you confirm you have this checked in your dev tools settings?
![image](https://github.com/phetsims/community/assets/6396244/c6acbe02-8517-411d-b41a-5c20f3848653)

**Reply by hiiambob89** [reply]
Thanks, I totally forgot about the cache.
